feat: Navigate prompt history in prompt field via arrow up/down (#4139)

hannesrudolph opened the pull request:
Fixes #4139

 ## Description

  This PR implements hybrid prompt history navigation using arrow up/down keys in the chat input field, providing an intuitive
  terminal-like experience that adapts to the user's current context.

  ### Key Features

  #### Hybrid Navigation Sources
  - **Active conversation**: Navigate through recent conversation messages (newest first)
  - **Starting fresh**: Navigate through task history from previous sessions (newest first)

  #### Smart Context Detection
  - Automatically detects if user is in an active conversation vs starting fresh
  - Seamlessly switches between `clineMessages` and `taskHistory` as data sources
  - Resets navigation position when switching between history sources

  #### Terminal-like UX
  - **Arrow Up**: Navigate to previous (older) prompts
  - **Arrow Down**: Navigate to next (newer) prompts or return to current input
  - Only triggers when cursor is at first/last line of textarea
  - Preserves current input when starting navigation
  - Restores current input when navigating back to present

  #### Performance & Memory Management
  - 100-item history limit to prevent memory issues
  - Efficient filtering of empty/whitespace-only messages
  - Workspace-aware filtering for task history

  ### Navigation Behavior

  1. **In Active Conversation**:
     - Press ↑ to get most recent conversation message
     - Continue ↑ to go through older conversation messages
     - Press ↓ to move forward through newer messages

  2. **Starting Fresh (No Active Conversation)**:
     - Press ↑ to get most recent task from history
     - Continue ↑ to go through older tasks
     - Press ↓ to move forward through newer tasks

  3. **Edge Cases**:
     - Empty conversation history: Navigation does nothing
     - Single prompt with no follow-ups: No navigation (fixed in latest commit)
     - Typing while navigating: Automatically resets navigation state

  ### Technical Implementation

  - Custom `usePromptHistory` hook encapsulates all navigation logic
  - Cursor position management for proper UX after navigation
  - Comprehensive test coverage for all navigation scenarios
  - Integration with existing chat component without breaking changes

hannesrudolph commented:
fixing issue where the history was inthe wrong order an it was also pulling from the pool of all history instead of just the workspace.

mochiya98 commented:
Thanks, it looks good!
There might be issues with `setTimeout(..., 0);` depending on the update timing, so it might be better to trigger it within `useLayoutEffect` .
However, since this affects state management more broadly, the implementation might be a bit more complex.

```ts
const [inputValue, setInputValue] = useState<{
  value: string;
  afterRender?: "SET_CURSOR_FIRST_LINE" | "SET_CURSOR_LAST_LINE";
}>({
  value: "",
});
useLayoutEffect(() => {
  if (!inputValue.afterRender) return;
  if (inputValue.afterRender === "SET_CURSOR_FIRST_LINE") {
    if (textAreaRef.current) {
      textAreaRef.current.setSelectionRange(0, 0);
    }
  } else if (inputValue.afterRender === "SET_CURSOR_LAST_LINE") {
    if (textAreaRef.current) {
      const lines = historicalPrompt.split("\n");
      const lastLineStart =
        historicalPrompt.length - lines[lines.length - 1].length;
      textAreaRef.current.setSelectionRange(lastLineStart, lastLineStart);
    }
  }
  setInputValue(({ value }) => ({ value }));
}, [inputValue]);

// Usage
setInputValue({
  value: "...",
  afterRender: "SET_CURSOR_FIRST_LINE",
});
setInputValue({
  value: "...",
  afterRender: "SET_CURSOR_LAST_LINE",
});
```

hannesrudolph commented:
## 🚀 Implemented Cursor Positioning Improvement

Thanks @mochiya98 for the excellent technical suggestion! I've implemented the `useLayoutEffect` approach you recommended to improve the cursor positioning reliability.

### Changes Made:

1. **Replaced `setTimeout(..., 0)` with `useLayoutEffect`**: More reliable timing for DOM manipulation
2. **Added `CursorPositionState` interface**: Better type safety for the state-based approach
3. **Implemented state-based cursor positioning**: Following the exact pattern you suggested

### Technical Details:

```typescript
interface CursorPositionState {
  value: string
  afterRender?: "SET_CURSOR_FIRST_LINE" | "SET_CURSOR_LAST_LINE" | "SET_CURSOR_START"
}
```

The new `useLayoutEffect` handles cursor positioning synchronously after DOM updates:
- `SET_CURSOR_FIRST_LINE`: Positions cursor at end of first line
- `SET_CURSOR_LAST_LINE`: Positions cursor at start of last line  
- `SET_CURSOR_START`: Positions cursor at beginning

### Benefits:
- ✅ More reliable cursor positioning timing
- ✅ Eliminates potential race conditions with `setTimeout`
- ✅ Better state management architecture
- ✅ All existing functionality preserved
- ✅ All 25 tests still passing

This improvement makes the prompt history navigation more robust, especially in scenarios where DOM updates might have variable timing. The cursor will now be positioned correctly regardless of render timing variations.

daniel-lxs commented on webview-ui/src/components/chat/ChatTextArea.tsx:
```diff
@@ -153,6 +159,30 @@ const ChatTextArea = forwardRef<HTMLTextAreaElement, ChatTextAreaProps>(
+				// - Arrow down decreases index to go forward (newer prompts)
+				setPromptHistory(prompts)
+			}
+		}, [taskHistory, cwd])
```
Should we consider adding a maximum history size limit to prevent memory issues with very long task histories?

daniel-lxs commented on webview-ui/src/components/chat/ChatTextArea.tsx:
```diff
@@ -153,6 +159,30 @@ const ChatTextArea = forwardRef<HTMLTextAreaElement, ChatTextAreaProps>(
+				// We keep it as-is so that navigation works correctly:
+				// - Arrow up increases index to go back in history (older prompts)
+				// - Arrow down decreases index to go forward (newer prompts)
+				setPromptHistory(prompts)
```
Consider memoizing the prompt history filtering to avoid recalculation on every render when `taskHistory` or `cwd` changes:

```typescript
const promptHistory = useMemo(() => {
  if (taskHistory && taskHistory.length > 0 && cwd) {
    return taskHistory
      .filter((item) => {
        return item.task && item.task.trim() !== "" && (!item.workspace || item.workspace === cwd)
      })
      .map((item) => item.task)
  }
  return []
}, [taskHistory, cwd])
```

This would improve performance for users with large task histories.

daniel-lxs commented on webview-ui/src/components/chat/ChatTextArea.tsx:
```diff
@@ -153,6 +159,30 @@ const ChatTextArea = forwardRef<HTMLTextAreaElement, ChatTextAreaProps>(
 		const [isFocused, setIsFocused] = useState(false)
 
+		// Prompt history navigation state
+		const [historyIndex, setHistoryIndex] = useState(-1)
```
Consider extracting the history navigation logic into a custom hook for better separation of concerns. This would reduce the complexity of the ChatTextArea component and make the history navigation logic more reusable and testable:

```typescript
const usePromptHistory = (taskHistory, cwd, inputValue) => {
  // Move historyIndex, tempInput, promptHistory state here
  // Move history navigation handlers here
  // Return { navigateHistory, resetHistory, promptHistory }
}
```

This would make the component easier to maintain and the history logic more focused.

daniel-lxs reviewed (commented):
Hey @hannesrudolph, I left some minor performance improvements suggestions.

mochiya98 commented on webview-ui/src/components/chat/ChatTextArea.tsx:
```diff
@@ -437,6 +462,27 @@ const ChatTextArea = forwardRef<HTMLTextAreaElement, ChatTextAreaProps>(
+				textAreaRef.current.setSelectionRange(0, 0)
+			}
+
+			setInputValueWithCursor({ value: inputValueWithCursor.value })
```
Thanks!
I'm the one who provided that part of the code snippet, but it should already be skipped in the process at Line 467 (without it, it would result in an infinite loop).
```ts
useLayoutEffect(() => {
  if (!inputValueWithCursor.afterRender || !textAreaRef.current) return
  // ...
```

daniel-lxs replied:
Thank you for clarifying this!

hannesrudolph commented:
## ✅ All Outstanding Review Concerns Addressed

I've completed implementing the enhanced hybrid prompt history navigation and have addressed all the review feedback. Here's a summary of what was addressed:

### 🔧 **Performance & Architecture (@daniel-lxs)**

1. **✅ Memory Management**: 
   - Added `MAX_PROMPT_HISTORY_SIZE = 100` limit to prevent memory issues
   - Implemented `.slice(-MAX_PROMPT_HISTORY_SIZE)` for both history sources

2. **✅ Performance Optimization**:
   - Replaced `useEffect` with `useMemo` for prompt history filtering
   - Prevents unnecessary recalculations on every render
   - Optimized dependency arrays

3. **✅ Code Organization**:
   - Extracted all history logic into `usePromptHistory.ts` custom hook
   - Improved separation of concerns and testability
   - Reduced complexity in `ChatTextArea.tsx`

4. **✅ Parameter Cleanup**:
   - Simplified `resetOnInputChange` callback (removed unused `_newValue` parameter)
   - Clean, focused function signatures

5. **✅ Defensive Programming**:
   - Added proper bounds checking for array access
   - Safe handling of edge cases and undefined states

6. **✅ History Navigation UX**:
   - Implemented **hybrid ordering** approach:
     - **Conversation messages**: Newest first (up arrow shows recent messages)
     - **Task history**: Oldest first (up arrow shows oldest tasks)
   - **Position reset** when switching between contexts
   - Terminal-like behavior that adapts to context

### 🎯 **Enhanced Implementation Beyond Original Scope**

- **Hybrid History Sources**: Automatically switches between conversation messages (`clineMessages`) and task history
- **Context-Aware Behavior**: Different sort orders for different use cases
- **Navigation Reset**: Fresh start when switching contexts
- **Comprehensive Testing**: Full test coverage for all scenarios
- **Backward Compatibility**: Maintains existing workflows

### 📝 **Updated Documentation**

The PR description has been completely updated to reflect the sophisticated hybrid implementation, including:
- Detailed technical explanations
- User experience examples
- Performance optimizations
- Testing coverage
- Backward compatibility assurance

---

**Ready for Review**: All feedback has been incorporated and the implementation goes beyond the original requirements to provide a more intuitive and performant user experience. 🚀

mrubens commented:
It feels unintuitive to me that the up arrow within a task can also bring up messages from other tasks. Is that part of this necessary, or could we just restrict it to navigating through prompts in the current task?

hannesrudolph commented:
> It feels unintuitive to me that the up arrow within a task can also bring up messages from other tasks. Is that part of this necessary, or could we just restrict it to navigating through prompts in the current task?

It should be restrcited to just the current task

hannesrudolph commented:
> It feels unintuitive to me that the up arrow within a task can also bring up messages from other tasks. Is that part of this necessary, or could we just restrict it to navigating through prompts in the current task?

Fix incoming. good catch.

daniel-lxs reviewed (approved):
Commit cc7bab0 solved the issue for me.

hannesrudolph commented:
@mrubens 
![image](https://github.com/user-attachments/assets/76281848-5a46-4874-8269-c004f5f80f66)
